vm/src/types.rs

```diff
@@ -9,36 +9,82 @@ pub type VMIndex = u32;
 pub type VMTag = u32;
 pub type VMInt = isize;
 
+/// Enum which represent the instructions executed by the virtual machine.
+///
+/// The binary arithmetic instructions pop two values of the stack and pushes the result of
+/// adding/comparing/etc the values. 
 #[derive(Copy, Clone, Debug)]
 pub enum Instruction {
+    /// Push an integer to the stack
     PushInt(isize),
+    /// Push a byte to the stack
     PushByte(u8),
+    /// Push a float to the stack
     PushFloat(f64),
+    /// Push a string to the stack by loading the string at `index` in the currently executing
+    /// function
     PushString(VMIndex),
+    /// Push a variable to the stack by loading the upvariable at `index` from the currently
+    /// executing function
+    PushUpVar(VMIndex),
+    /// Push the value at `index`
     Push(VMIndex),
+    /// Push the value at `index`
     PushGlobal(VMIndex),
+    /// Call a function by passing it `args` number of arguments. The function is at the index in
+    /// the stack just before the arguments. After the call is done the function all arguments are
```

Comment:
Remove 'the function' here - 'After the call is done all arguments are removed and the function is replaced by the result of the call.'